src/assets/scripts/client/trafficGenerator/buildPreSpawnAircraft.js

```diff
@@ -1,78 +1,245 @@
+import _findIndex from 'lodash/findIndex';
+import _forEach from 'lodash/forEach';
 import _isNil from 'lodash/isNil';
+import _random from 'lodash/random';
 import RouteModel from '../aircraft/FlightManagementSystem/RouteModel';
-// import { routeStringFormatHelper } from '../navigationLibrary/Route/routeStringFormatHelper';
 import {
     isWithinAirspace,
     calculateDistanceToBoundary
 } from '../math/flightMath';
 import { nm } from '../utilities/unitConverters';
 import { isEmptyObject } from '../utilities/validatorUtilities';
 
+/**
+ * Return an array whose indices directly mirror those of `waypointModelList`, except it
+ * contains the distances along the route at which each waypoint lies from the spawn point
+ *
+ * @function _calculateOffsetsToEachWaypointInRoute
+ * @param waypointModelList {array<WaypointModel>} array of all waypoints along the route
+ * @return {array<number>} offset distance from spawn point to each waypoint
+ */
+function _calculateOffsetsToEachWaypointInRoute(waypointModelList) {
+    // begin by storing the first fix's offset: a distance of 0 from the spawn point
+    const waypointOffsetMap = [0];
+
+    // continue with second waypoint, because first is already stored
+    for (let i = 1; i < waypointModelList.length; i++) {
+        if (waypointModelList[i].isVectorWaypoint || waypointModelList[i - 1].isVectorWaypoint) {
+            continue;
+        }
+
+        const previousWaypointModel = waypointModelList[i - 1];
+        const nextWaypointModel = waypointModelList[i];
+        const distanceToNextWaypoint = previousWaypointModel.calculateDistanceToWaypoint(nextWaypointModel);
+
+        waypointOffsetMap.push(distanceToNextWaypoint);
+    }
+
+    return waypointOffsetMap;
+}
+
+/**
+ * Create an array of the various altitude restrictions along the route, and the offset distances at which they exist
+ *
+ * Note that unlike the `waypointOffsetMap`, the altitude offset array DOES NOT mirror the indices
+ * of the `waypointModelList`. Only offsets and altitudes of altitude-restricted waypoints are included.
+ *
+ * @function _calculateAltitudeOffsets
+ * @param waypointModelList {array<WaypointModel>} array of all waypoints along the route
+ * @param waypointOffsetMap {array<number>} offset distance from spawn point to each waypoint
+ * @return {array<array<number>>} [[offsetDistance, altitude], [offsetDistance, altitude], ...]
+ */
+function _calculateAltitudeOffsets(waypointModelList, waypointOffsetMap) {
+    const altitudeOffsets = [];
+
+    _forEach(waypointModelList, (waypointModel, waypointIndex) => {
```

Comment:
If `waypointModelList` is actually an array, we should use a simple for loop here instead.

Reply:
done